modules/data_warehouse/main.tf

```diff
@@ -0,0 +1,396 @@
+  location       = var.region
+  data_source_id = "scheduled_query"
+  schedule       = "every day 00:00"
+  params = {
+    query = "CALL `${var.project_id}.ds_edw.sp_lookerstudio_report`()"
+  }
+
+  depends_on = [
+    google_project_iam_member.dts_permissions_token,
+    google_project_iam_member.dts_permissions_agent,
+    google_bigquery_dataset.ds_edw
+  ]
+}
+
+# Create a Cloud Function resource
+# # Zip the function file
+data "archive_file" "bigquery_external_function_zip" {
+  type        = "zip"
+  source_dir  = "${path.module}/assets/bigquery-external-function"
+  output_path = "${path.module}/assets/bigquery-external-function.zip"
+
+  depends_on = [
+    google_storage_bucket.provisioning_bucket
+  ]
+}
+
+# # Place the function file on Cloud Storage
+resource "google_storage_bucket_object" "cloud_function_zip_upload" {
+  name   = "assets/bigquery-external-function.zip"
+  bucket = google_storage_bucket.provisioning_bucket.name
+  source = data.archive_file.bigquery_external_function_zip.output_path
+
+  depends_on = [
+    google_storage_bucket.provisioning_bucket,
+    data.archive_file.bigquery_external_function_zip
+  ]
+}
+
+# # Create the function
+resource "google_cloudfunctions2_function" "function" {
+  #provider = google-beta
+  project     = var.project_id
+  name        = "bq-sp-transform-${random_id.id.hex}"
+  location    = var.region
+  description = "gcs-load-bq"
+
+  build_config {
+    runtime     = "python310"
+    entry_point = "bq_sp_transform"
+    source {
+      storage_source {
+        bucket = google_storage_bucket.provisioning_bucket.name
+        object = "assets/bigquery-external-function.zip"
+      }
+    }
+  }
+
+  service_config {
+    max_instance_count = 1
+    available_memory   = "256M"
+    timeout_seconds    = 540
+    environment_variables = {
+      PROJECT_ID = var.project_id
+      BUCKET_ID  = google_storage_bucket.raw_bucket.name
+    }
+    service_account_email = google_service_account.cloud_function_service_account.email
+  }
+
+  event_trigger {
+    trigger_region = var.region
+    event_type     = "google.cloud.storage.object.v1.finalized"
+    event_filters {
+      attribute = "bucket"
+      value     = google_storage_bucket.raw_bucket.name
+    }
+    retry_policy = "RETRY_POLICY_RETRY"
+  }
+
+  depends_on = [
+    google_storage_bucket.provisioning_bucket,
+    google_storage_bucket.raw_bucket,
+    google_project_iam_member.cloud_function_service_account_editor_role
+  ]
+}
+
+resource "google_project_iam_member" "workflow_event_receiver" {
+  project = var.project_id
+  role    = "roles/eventarc.eventReceiver"
+  member  = "serviceAccount:${data.google_project.project.number}-compute@developer.gserviceaccount.com"
+}
+
+resource "google_storage_bucket_object" "startfile" {
+  bucket = google_storage_bucket.raw_bucket.name
+  name   = "startfile"
+  source = "${path.module}/assets/startfile"
+
+  depends_on = [
+    google_cloudfunctions2_function.function
```

Comment:
nit: Would we need some kind of time buffer before `google_cloudfunctions2_function.function` is ready to work?

Reply:
i found that the depends_on worked properly but we can validate this with future deployments.